File: ts/2025/day07.ts
let p1 = 0, p2 = 0

const lines = (await Bun.stdin.text()).trim().split("\n").map(line => line.split(""))
let ways = new Map<string, number>()

const n = lines.length
for (const [row, line] of lines.slice(0, -1).entries()) {
	for (const [col, ch] of line.entries()) {
		if (ch === "|" || ch === "S") {
			let nextRow = lines[row + 1]
			if (nextRow === undefined) { continue }
			let cnt = ways.get(`${row},${col}`) ?? 1
			const char = nextRow[col]
			if (char === "^") {
				nextRow[col - 1] = "|"
				ways.set(`${row + 1},${col - 1}`, cnt + (ways.get(`${row + 1},${col - 1}`) ?? 0))
				nextRow[col + 1] = "|"
				ways.set(`${row + 1},${col + 1}`, cnt + (ways.get(`${row + 1},${col + 1}`) ?? 0))
				p1++
			}
			if (char === "." || char === "|") {
				nextRow[col] = "|"
				ways.set(`${row + 1},${col}`, cnt + (ways.get(`${row + 1},${col}`) ?? 0))
			}
		}
	}
}
for (const [col, ch] of lines.at(n - 1)!.entries()) {
	if (ch === "|") { p2 += (ways.get(`${n - 1},${col}`) ?? 0) }
}
console.log("Part 1: %d", p1)
console.log("Part 2: %d", p2)

File: py/2025/day07.py
import sys
from itertools import product
from collections import defaultdict

p1, p2 = 0, 0
D = sys.stdin.read()
lines = [list(line) for line in D.split("\n") if line]

M = defaultdict(int)
n = len(lines)
for row, col in product(range(n - 1), range(n - 1)):
    ch = lines[row][col]
    if ch == "|" or ch == "S":
        cnt = M.get((row, col), 1)
        char = lines[row + 1][col]
        if char == "^":
            lines[row + 1][col - 1] = "|"
            M[(row + 1, col - 1)] += cnt
            lines[row + 1][col + 1] = "|"
            M[(row + 1, col + 1)] += cnt
            p1 += 1
        if char == "." or char == "|":
            lines[row + 1][col] = "|"
            M[(row + 1, col)] += cnt

r = len(lines) - 1
p2 = sum(M.get((r, col), 0) for col, ch in enumerate(lines[-1]) if ch == "|")

print(f"The result for part 1: {p1}")
print(f"The result for part 2: {p2}")
